File: implementations/python/src/foilbench_py/solvers/factory.py
"""Solver registry."""

from foilbench_py.core.protocol import FlowSolver
from foilbench_py.solvers.lbm import LBMSolver
from foilbench_py.solvers.pic_flip import PicFlipSolver
from foilbench_py.solvers.stable_fluids import StableFluidsSolver

_SOLVERS = ("stable-fluids", "lbm-d2q9", "pic-flip")


def solver_ids() -> tuple[str, ...]:
    return _SOLVERS


def create_solver(solver_id: str) -> FlowSolver:
    if solver_id == "stable-fluids":
        return StableFluidsSolver()
    if solver_id == "lbm-d2q9":
        return LBMSolver()
    if solver_id == "pic-flip":
        return PicFlipSolver()
    raise KeyError(f"unknown solver {solver_id!r}; choose from {', '.join(_SOLVERS)}")

File: implementations/typescript/src/solvers/factory.ts
import type {FlowSolver, SolverId} from "../core/contracts.js";
import {StableFluidsSolver} from "./stableFluids.js";
import {LbmSolver} from "./lbm.js";
import {PicFlipSolver} from "./picFlip.js";

export function createSolver(id: SolverId): FlowSolver {
  switch (id) {
    case "stable-fluids": return new StableFluidsSolver();
    case "lbm-d2q9": return new LbmSolver();
    case "pic-flip": return new PicFlipSolver();
    default: return unreachable(id);
  }
}

function unreachable(value: never): never {
  throw new Error(`unsupported solver id: ${String(value)}`);
}
